bin/setupReplication.js

```diff
@@ -0,0 +1,431 @@
+        this._targetBucket = targetBucket;
+        this._s3Clients = {
+            source: _setupS3Client(source.s3.host, source.s3.port,
+                'backbeatsource'),
+            target: _setupS3Client(destination.s3.host, destination.s3.port,
+                'backbeattarget'),
+        };
+        this._iamClients = {
+            source: _setupIAMClient(source.auth.vault.host,
+                 source.auth.vault.iamPort,
+                'backbeatsource'),
+            target: _setupIAMClient(destination.auth.vault.host,
+                 destination.auth.vault.iamPort,
+                'backbeattarget'),
+        };
+    }
+
+    _checkSanity(cb) {
+        return async.waterfall([
+            next => this._isValidBucket('source', next),
+            next => this._isValidBucket('target', next),
+            next => this._isVersioningEnabled('source', next),
+            next => this._isVersioningEnabled('target', next),
+            next => this._isReplicationEnabled('source', next),
+            (srcArn, next) => this._isValidRole('source', srcArn, next),
+            (tgtArn, next) => this._isValidRole('target', tgtArn, next),
+        ], cb);
+    }
+
+    _isValidBucket(where, cb) {
+        // Does the bucket exist and is it reachable?
+        const bucket = where === 'source' ? this._sourceBucket :
+            this._targetBucket;
+        this._s3Clients[where].headBucket({ Bucket: bucket }, err => {
+            if (err) {
+                this._log.error('bucket sanity check error', {
+                    error: err,
+                    method: '_SetupReplication._isValidBucket',
+                    errorMessage: err.message,
+                    errorStack: err.stack,
+                });
+                return cb(err);
+            }
+            return cb();
+        });
+    }
+
+    _isVersioningEnabled(where, cb) {
+        // Does the bucket have versioning enabled?
+        const bucket = where === 'source' ? this._sourceBucket :
+            this._targetBucket;
+        this._s3Clients[where].getBucketVersioning({ Bucket: bucket },
+            (err, res) => {
+                if (err || res.Status === 'Disabled') {
+                    this._log.error('versioning sanity check error', {
+                        error: err,
+                        method: '_SetupReplication._isVersioningEnabled',
+                        errorMessage: err.message,
+                        errorStack: err.stack,
+                    });
+                    return cb(err);
+                }
+                return cb();
+            }
+        );
+    }
+
+    _isValidRole(where, arn, cb) {
+        // Is the role mentioned in the replication config available in IAM
+
+        // Goal is to get Role given known ARN.
+        // If err, there is no matching role
+        let roleName;
+        let tgtArn;
+
+        if (where === 'source') {
+            [roleName, tgtArn] = arn.split(',');
```

Comment:
Rather than doing this parsing here, I would rather put it outside `_isValidRole`, in the main waterfall block after `_getBucketReplication`, and calling twice `_isValidRole` then with source and target parts of the comma-separated roles. I know this could spoil the nice layout of function calls in the waterfall :face_with_head_bandage: but it's better IMO to have `_isValidRole` always taking the same kind of `arn` argument which is a single role arn, rather than a different one if source or target.